lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -823,6 +846,90 @@ data ByronWalletPutPassphraseData = ByronWalletPutPassphraseData
     , newPassphrase :: !(ApiT (Passphrase "raw"))
     } deriving (Eq, Generic, Show)
 
+-- | Polymorphic wrapper for byte arrays, parameterised by the desired string
+-- encoding.
+newtype ApiBytesT (base :: Base) bs = ApiBytesT { getApiBytesT :: bs }
+    deriving (Generic, Show, Eq, Functor)
+    deriving newtype (Semigroup, Monoid, Hashable)
+    deriving anyclass NFData
+
+newtype ApiSerialisedTransaction
+    = ApiSerialisedTransaction (ApiBytesT 'Base64 SerialisedTx)
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+data ApiConstructTransaction (n :: NetworkDiscriminant) = ApiConstructTransaction
+    { serializedTransaction :: !ApiSerialisedTransaction
+    , coinSelection :: !(ApiCoinSelection n)
+    , fee :: !(Quantity "lovelace" Natural)
+    } deriving (Eq, Generic, Show)
+      deriving anyclass NFData
+
+-- | Index of the stake key.
+newtype StakeKeyIndex = StakeKeyIndex (ApiT DerivationIndex)
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+-- | Stake pool delegation certificates.
+data ApiMultiDelegationAction
+    = Joining !(ApiT PoolId) !StakeKeyIndex
+    -- ^ Delegate given staking index to a pool, possibly registering the stake
+    -- key at the same time.
+    | Leaving !StakeKeyIndex
+    -- ^ Undelegate the given staking index from its pool.
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+-- | Input parameters for transaction construction.
+data ApiConstructTransactionData (n :: NetworkDiscriminant) = ApiConstructTransactionData
+    { payments :: !(Maybe (ApiPaymentDestination n))
+    , withdrawal :: !(Maybe ApiWithdrawalPostData)
+    , metadata :: !(Maybe (ApiT TxMetadata))
+    , mint :: !(Maybe (ApiT W.TokenMap))
+    , delegations :: !(Maybe [ApiMultiDelegationAction])
```

Comment:
I forget, why did we have `Maybe [ApiMultiDelegationAction]` here?
Was there a semantic difference between `Nothing` and `Just []`?

Reply:
It is reasonable to use `delegations: Maybe ...` as it can be optional on Api level, one can omit "delegations" field.. I will use `Maybe (NonEmpty ApiMultiDelegationAction)` though

Reply:
Aside from arguably looking a bit strange, wouldn't `Maybe (NonEmpty ApiMultiDelegationAction)` forbid `delegations = []`, potentially making it more difficult for folks to write API clients?

Could we not have `delegations :: [ApiMultiDelegationAction]`, but have a custom JSON parser defaulting to `[]` if the key is not present?